### PR TITLE
Fix new lines into empty implementation

### LINKED ISSUE #5060: Formatter adds newline inside empty impl block
After running the formatter:

```
-impl OrdEq for u64 {}
-impl OrdEq for u32 {}
-impl OrdEq for u16 {}
-impl OrdEq for u8 {}
-impl OrdEq for b256 {}
+impl OrdEq for u64 {
+}
+impl OrdEq for u32 {
+}
+impl OrdEq for u16 {
+}
+impl OrdEq for u8 {
+}
+impl OrdEq for b256 {
+}
```

Related https://github.com/FuelLabs/sway/issues/5052

### PR DESCRIPTION
**Pull request by cr-fuel**
Fixes #5060 

## Description

Empty implementations have a bug with new lines

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: swayfmt/src/items/item_impl/mod.rs [hunk 1]
```diff
@@ -38,14 +38,27 @@ impl Format for ItemImpl {
...
-            item.format(formatted_code, formatter)?;
-            writeln!(formatted_code)?;
+        if contents.is_empty() {
+            Self::open_curly_brace(formatted_code, formatter)?;
+            let comments = write_comments(formatted_code, self.span().into(), formatter)?;
+            if !comments {
+                formatter.unindent();
+                if formatted_code.ends_with('\n') {
+                    formatted_code.pop();
+                }
```
Comment (eureka-cpu) [inline swayfmt/src/items/item_impl/mod.rs:50]: Please do not use `pop()`, or other destructive methods if possible, especially on the returned formatted string.

Comment (cr-fuel) [reply swayfmt/src/items/item_impl/mod.rs]: My approach was tu use `truncate` but I need to make the current string smaller. I'll check #5149 to check how you did it.

File: swayfmt/src/constants.rs [hunk 2]
```diff
@@ -48,6 +48,11 @@ pub(crate) const CARRIAGE_RETURN: char = '\r';
 pub(crate) const WINDOWS_NEWLINE: &str = "\r\n";
 pub(crate) const UNIX_NEWLINE: &str = "\n";
 
+#[cfg(target_os = "windows")]
```
Comment (JoshuaBatty) [inline swayfmt/src/constants.rs:51]: These look like they are added also in #5178

Comment (cr-fuel) [reply swayfmt/src/constants.rs:51]: @JoshuaBatty yes, the first PR to be merged wins. The other will have a smaller diff.

File: swayfmt/src/items/item_impl/mod.rs (hunk 1 above)
Comment (cr-fuel) [reply swayfmt/src/items/item_impl/mod.rs]: @eureka-cpu I removed the destructive method in a8b8e0379

File: swayfmt/src/items/item_fn/mod.rs [hunk 3]
```diff
@@ -51,11 +51,11 @@ impl Format for ItemFn {
...
                 } else {
+                    let range = self.span().into();
                     Self::open_curly_brace(formatted_code, formatter)?;
-                    formatter.indent();
-                    let comments = write_comments(formatted_code, self.span().into(), formatter)?;
-                    if !comments {
-                        formatter.unindent();
+                    if has_comments_in_formatter(formatter, &range) {
+                        formatter.indent();
+                        write_comments(formatted_code, range, formatter)?;
```
Comment (sdankel) [inline swayfmt/src/items/item_fn/mod.rs:58]: Does this change the behavior of this block or just make it more readable?

I believe this pattern of checking `!comments` is used in a bunch of other parts of the formatter. We might want to use it everywhere to establish it as a pattern.

**Review by sdankel — APPROVED** [review]

File: swayfmt/src/items/item_fn/mod.rs (hunk 3 above)
Comment (cr-fuel) [reply swayfmt/src/items/item_fn/mod.rs:58]: If we check it after we would have to remove the bits from the strings (the newline that was added). If we checked this way we prevent adding it, this kind of things was introduced in #5148 (which is nicer I think)

**Review by JoshuaBatty — APPROVED** [review]

File: swayfmt/src/comments.rs [hunk 4]
```diff
@@ -35,6 +35,18 @@ impl CommentsContext {
...
+pub fn has_comments_in_formatter(formatter: &Formatter, range: &Range<usize>) -> bool {
+    formatter
+        .comments_context
+        .map
+        .comments_between(range)
+        .peekable()
+        .peek()
+        .is_some()
+}
+
```
Comment (eureka-cpu) [inline swayfmt/src/comments.rs:48]: Nice one, very clean

**Review by eureka-cpu — APPROVED** [review]
Much better, nice work!

**Pull request merged**